sycl/unittests/pi/EnqueueMemTest.cpp

```diff
@@ -25,72 +23,77 @@ class DISABLED_EnqueueMemTest : public ::testing::Test {
   pi_queue _queue = nullptr;
   pi_mem _mem = nullptr;
 
-  DISABLED_EnqueueMemTest() = default;
+  EnqueueMemTest() = default;
 
-  ~DISABLED_EnqueueMemTest() = default;
+  ~EnqueueMemTest() = default;
 
   void SetUp() override {
-    Plugins = detail::pi::initialize();
-    ASSERT_FALSE(Plugins.empty());
+
+    detail::plugin plugin = GetParam();
 
     pi_platform platform = nullptr;
-    ASSERT_EQ((Plugins[0].call_nocheck<detail::PiApiKind::piPlatformsGet>(
+    ASSERT_EQ((plugin.call_nocheck<detail::PiApiKind::piPlatformsGet>(
                   1, &platform, nullptr)),
               PI_SUCCESS);
 
-    ASSERT_EQ((Plugins[0].call_nocheck<detail::PiApiKind::piDevicesGet>(
-                  platform, PI_DEVICE_TYPE_GPU, 1, &_device, nullptr)),
+    ASSERT_EQ((plugin.call_nocheck<detail::PiApiKind::piDevicesGet>(
+                  platform, PI_DEVICE_TYPE_DEFAULT, 1, &_device, nullptr)),
```

Comment:
What is the expected behavior here? Does it match documented OpenCL one?

Reply:
Correct.

Reply:
This is what OpenCL does:
~~~
CL_DEVICE_TYPE_DEFAULT | The default OpenCL device in the system.
CL_DEVICE_TYPE_ALL | All OpenCL devices available in the system.
~~~
I don't see how they nicely translate to PI. What is the semantics of `PI_DEVICE_TYPE_DEFAULT` and 'PI_DEVICE_TYPE_DEFAULT' that plugins must implement? How is the "default" device identified? Does "all" mean all under this plugin only? And why did you need to add these in the first place?

Reply:
These enumerators are implicitly converted to the matching OpenCL enumerator. This means that there is no work required to implement these in OpenCL. They are simply casted to `CL_DEVICE_TYPE_*` and passed to `clGetDeviceIDs` as they would normally be used. The work has been done to enable these new enumerators for CUDA in this pull request.

It is necessary to expose `CL_DEVICE_TYPE_DEFAULT` on the PI API because I am testing multiple plugins in the unit tests now. If I hard code a specific device type, such as a GPU, I may fall upon a plugin that does not have that specific type of device. This was the case when I first ran the tests, as the OpenCL on my system does not have a GPU device, it hits a runtime assertion within PI OpenCL's `piDevicesGet` implementation and makes the test fall over. By exposing the idea of a "default" device, I can simply request each backend pass me back a device that will work. For OpenCL plugins, the default PI device will be whatever they would return when querying `clGetDeviceIDs` with `CL_DEVICE_TYPE_DEFAULT`. For CUDA, the default pi device is a GPU. Being able to ask PI API for a device to work on without hard-coding the device type is essential to the work I am doing here, and previously there was not a way to do this without implicit conversions from OpenCL or SYCL device type enumerators. 

There are places in the runtime that rely on implicit conversions from enumerators that have no equivelant in the PI API header. An example of this is `CL_DEVICE_TYPE_ALL`. The sycl headers define `device_type::all` as having the value `CL_DEVICE_TYPE_ALL`. There is no equivelant to this enumerator in PI API. From the perspecive of PI API, you cannot query for "all" devices. There is nothing in pi.h that documents this functionality. And yet this value is passed to PI API all the time, because `device_type::all` is the default value when you attempt to get devices in SYCL. It works right now because it casts SYCL's `device_type` enum class to PI API's pi_device_type enum, which is implicitly convertable to OpenCL's `cl_device_type` enum. The SYCL runtime is essentially relying on a hidden feature. I have tried to legitimise it by making PI_DEVICE_TYPE_ALL a proper enumerator. I do not need PI_DEVICE_TYPE_ALL in this PR, so I can split this out and make it a seperate PR if you would prefer.

Reply:
>There are places in the runtime that rely on implicit conversions from enumerators that have no equivalent in the PI API header.

To be clear, I am totally supporting population of PI API with needed values and eventually being able to build SYCL RT without OpenCL at all.

> By exposing the idea of a "default" device, I can simply request each backend pass me back a device that will work. 

Thanks for explaining this, and I agree this is a useful semantics.
Since this is not 1:1 maps OpenCL description of `CL_DEVICE_TYPE_DEFAULT` can you please put a word about this PI meaning into pi.h?

> And yet this value is passed to PI API all the time, because device_type::all is the default value when you attempt to get devices in SYCL.

And then I assume we should document this as returning all devices available under a particular plugin, correct? I am fine adding `PI_DEVICE_TYPE_ALL` but, again, please document what it is returning in pi.h

Reply:
No problem, thanks for the discussion. I am glad we agree that decoupling SYCL RT from OpenCL is important. I think you are right that making clear how these values differ from their OpenCL equivalents will be useful, so I will update the PR to document all the pi_device_type enumerators.

Reply:
I have documented the enumerators.